### PR TITLE
Cache and doc changes

### PR DESCRIPTION
**Pull request by Sorck**
I've added a file based cache which works similarly to SMF 2.x
Cache calls are now all type checks so as to prevent problems
DB dump has been included. You might want to discard it or put it in another repo just so, as Frustrate said, 'try and use it'. :P
I've made a few other changes, see the commits for a little more info.

Sorry about some things having terrible commit names, wrong author name and not being signed off properly. Even when using Netbeans Git can confuse me. -_- But I should've worked out how to use Git now!

I was thinking of making all the cache drivers return null on failure (or throwing an exception) as that implies no data having been stored. Returning false is a little silly as it might have wanted to return a stored bool.

### CODE REVIEW

**Comment by Fustrate** [comment]
I'll look over this in a few hours, have to study for tonight's exam. Might have to alter some of the code because of the changes I've made in the past few days, sorry about that. Pre-alpha software and all :)

**Comment by Sorck** [comment]
Yeah, I noticed the gitignore file and understood why that was there. 
I think I fixed it in a later commit.

**Comment by Fustrate** [comment]
I don't agree with the cache checks being `is_array()` - if you want to store `false`, store it as `array(false)` and then check it. There's no value in PHP that won't cause an issue like this, and having to separate it into two lines doesn't feel right.

**Comment by Sorck** [comment]
No problem with code changes - it is an communal effort!

Maybe we could force an array type check then for the cache save function?

**Comment by Fustrate** [comment]
No, we're not going to store everything as an array - that's restricting the component too much. Let's leave the method signature alone for now, it shouldn't cause an issue if used properly.

**Comment by Fustrate** [comment]
> No problem with code changes - it is an communal effort!

It's nice to have someone interested in contributing to the code so early, by the way :) it's been very lonely so far...

**Comment by Sorck** [comment]
OK. I'll just put a line or two in the File cache driver docs to say about it returning false whether if it's non existent or a bool has been stored.

**Comment by Sorck** [comment]
If a few more people contribute regularly then smCore might be viable for a beta in a couple of years. :P

Yep. smCore is fairly quiet at the moment.

**Comment by Fustrate** [comment]
Here's a secret: there's a full website out in the wild running a modified version of smCore ;) granted there are lots of hard-coded things that would need to be changed for a distributable version of smCore, but it's been a fun project so far.

**Comment by Sorck** [comment]
I think I'll try and port SimpleWiki over to smCore when it's working reasonably (e.g. when a working database dump is ready).

Then there could be another website running smCore! :P

**Comment by Fustrate** [comment]
I'm still working on an auth module, that's one of the biggest things we currently need and would be important to porting SimpleWiki. After that... I'm thinking parsers (BBCode, Markdown, etc.).

**Comment by Sorck** [comment]
Yeah. IIRC the SMF BBCode parser, although being fantastic, was considered slow? It might need a rewrite.

I don't think I've ever seen one which works as well though (as in doesn't make nesting errors).

Auth is probably one of the major things to do. Can't have a dynamic website without it.

File: other/database.sql [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+  PRIMARY KEY (`id_route`)
+) ENGINE=MyISAM DEFAULT CHARSET=utf8 AUTO_INCREMENT=1 ;
+
+-- --------------------------------------------------------
+
+--
+-- Table structure for table `sessions`
+--
+
+CREATE TABLE IF NOT EXISTS `sessions` (
```
Comment (Fustrate) [inline other/database.sql:202]: Normalized column names: `id_session`, `session_expires`, `session_data`

File: other/database.sql [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+-- --------------------------------------------------------
+
+--
+-- Table structure for table `lang_strings`
+--
+
+CREATE TABLE IF NOT EXISTS `lang_strings` (
+  `string_key` varchar(32) NOT NULL,
+  `string_value` varchar(512) NOT NULL,
+  `id_package` int(11) NOT NULL
```
Comment (Fustrate) [inline other/database.sql]:
``` sql
CREATE TABLE IF NOT EXISTS `lang_strings` (
  `id_string` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `string_language` smallint(4) unsigned NOT NULL,
  `string_package` smallint(4) unsigned NOT NULL,
  `string_key` varchar(255) CHARACTER SET ascii NOT NULL DEFAULT '',
  `string_value` text NOT NULL,
  PRIMARY KEY (`id_string`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

File: .htaccess [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# the following two protect our settings files
+RedirectMatch 403 ^.*/settings.php
+RedirectMatch 403 ^.*/other/*
```
Comment (Fustrate) [inline .htaccess]: This line can be removed, the /other/ directory won't be part of an official install.

File: library/smCore/Cache/Driver/File.php [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+		return;
+	}
+
+	/**
+	 * this function makes calculating the key hash easier
+	 * 
+	 * @param string $key
+	 * @return string-32
+	 */
+	protected function makeKey($key) {
```
Comment (Fustrate) [inline library/smCore/Cache/Driver/File.php]: This brace belongs on the next line, Allman style.

File: library/smCore/Cache/Driver/File.php [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	{
+		// not sure on this either
+		return;
+	}
+
+	/**
+	 * this function makes calculating the key hash easier
+	 * 
+	 * @param string $key
+	 * @return string-32
```
Comment (Fustrate) [inline library/smCore/Cache/Driver/File.php]: We're not going to add type lengths to `param` or `return` types.

File: .htaccess (hunk 3 above)
Comment (Sorck) [inline .htaccess]: Fixed in latest commit.

File: other/database.sql (hunk 2 above)
Comment (Sorck) [inline other/database.sql]: Fixed in commit.

File: other/database.sql (hunk 1 above)
Comment (Sorck) [inline other/database.sql:202]: Fixed in commit.

**Comment by Sorck** [comment]
I believe every file I've touched should work now.

Though I think only the settings, database dump and File cache driver need merging in the main repo. Not sure if there's a way to specifically do this or not. :P

And once this done I'll have to work out how to make my repository an exact copy of the 'official' one... maybe just re-pulling it?

**Comment by Fustrate** [comment]
Looks like you're going to have to merge the changes from the SimpleMachines repo into your copy before this can be merged. I think `git pull` accomplishes that, but I'm not sure what you're using to do your git stuff. I'm still learning all of this myself :)

**Comment by Sorck** [comment]
I'm running Netbeans under x64 Windows 7.

Seems like an upstream pull is needed but I'm clueless how to do it inside of Netbeans. Might just manually merge as the major changes I made were adding new files.

**Comment by Sorck** [comment]
Merge complete and some syntax errors corrected.

The SQL dump needs updating though with the thus far non-existent 'official' version.

**Comment by Fustrate** [comment]
One last thing and it'll be ready - \smCore\Application and \smCore\Model\Language have windows EOL characters (`\r`/`CR`) which need to be replaced with Unix EOL characters (`\n`/`LF`). Netbeans should force all files to `LF` in the future, since I just added a `.gitattributes` file last night, so hopefully you won't have to worry about it again.

**Comment by Fustrate** [comment]
(After this gets merged, just do a pull from the SimpleMachines repo to get the .gitattributes file, or you can add it directly from https://github.com/SimpleMachines/smCore/blob/master/.gitattributes)

**Comment by Sorck** [comment]
I would think I've fixed the EOL errors.

Not sure how I could verify that though.

**Comment by Fustrate** [comment]
Actually, Application is fine now. `/Storage/Modules.php` and `/Model/Language.php` still look like they're using carriage returns, though. You can tell because in the diff view on github, the whole file is being replaced instead of individual portions.

**Comment by Fustrate** [comment]
Also a note, if you use IRC I'm usually online most of the day (and some of the night, too). #smf and #smcore on Freenode.

**Comment by Sorck** [comment]
I'm going for the most commits added to a pull request...

All files you've mentioned have now been fixed I believe.

**Comment by Fustrate** [comment]
Modules Storage looks good, just the Language Model now

**Comment by Sorck** [comment]
I have tried to replace that several times with no success. :-\

**Comment by Fustrate** [comment]
http://fustrate.com/Language.php

That should work. There was also an extra blank line added before the opening `<?php` which I removed.

**Comment by Sorck** [comment]
The version in my repo has now lost that excess line.

**Comment by Sorck** [comment]
Fixed now... I think.
It really shouldn't have taken that much effort though. I must remember that editing in Notepad++ will solve EOL issues.

**Comment by Fustrate** [comment]
Once you have the .gitattributes file, it shouldn't be a problem anymore (as long as Netbeans honors it, which it should).

**Comment by Sorck** [comment]
As I'm sure everyone knows, just because it should doesn't mean it does.

**Pull request merged**